### ISSUE TITLE
CI issues

### ISSUE THREAD

**Issue by markus2330** [ElektraInitiative/PermaplanT#241]
- [ ] PR deployment seems to be broken (#226 was rebuilt but still old page was on pr.permaplant.net)
- [ ] `cargo test` rebuilds many deps, please add this to Docker container
- [ ] build time is already quite high: maybe run "Checks" and "Build in Docker" in parallel?

**Comment by lukashartl** [comment]
> PR deployment seems to be broken (https://github.com/ElektraInitiative/PermaplanT/pull/226 was rebuilt but still old page was on pr.permaplant.net)

If I look into the pipeline log, I can see that the frontend files and the backend binary were successfully copied into the right destination. Service restart also occurred. So it's really hard for me to reconstruct this. It could also be a caching issue. Maybe we should take a closer look into caching, caching headers, configuration for Vite, etc.?

If we had something like #164 it would be also easier to debug this.

For the time being I'll write a txt file which contains the currently deployed SHA-Commit into the web-root.

I'll discuss the other two points with my buddy to see how we can optimize building and testing Rust.

**Comment by markus2330** [comment]
> If I look into the pipeline log, I can see that the frontend files and the backend binary were successfully copied into the right destination. Service restart also occurred. So it's really hard for me to reconstruct this.

Yes, of course. It is better to keep an eye on it in new PRs. I just noted it so that we can have an eye on it. It is possible that there was another build job directly afterwards.

> It could also be a caching issue. Maybe we should take a closer look into caching, caching headers, configuration for Vite, etc.?

Yes, good idea. I cannot exclude this possibility, as this is really a quite complex topic.

> If we had something like https://github.com/ElektraInitiative/PermaplanT/issues/164 it would be also easier to debug this.

It was very obvious that it was not the PR in question.

> For the time being I'll write a txt file which contains the currently deployed SHA-Commit into the web-root.

Yes, simple and good solution :+1: Then I can compare with the Jenkins build log (at least the case that another PR is deployed directly afterwards should be clearly visible.)

> I'll discuss the other two points with my buddy to see how we can optimize building and testing Rust.

It doesn't have a priority, I simply noted it to not forget. And such things easily accumulate and suddenly we would have a  bad build time :wink: At the moment it is okay :rocket:

**Comment by markus2330** [comment]
Btw. also some report which npm, vite version etc. was used in the web root might be useful from time to time. It *also* should be in the build log.

**Comment by lukashartl** [comment]
Ok, so I did two changes for now:

1. I added the `Cache-Control: no-cache` header. (this should force the use of the ETag header instead of going directly into the cache)
2. I added the `.commit` file, which looks like this:

```bash
$ curl https://pr.permaplant.net/.commit
35f3430 (ci: DEPLOY_PROD also on non-master-branches [#79], 2023-04-05)
```

Unfortunately, if I want to open it in the browser (after I have already opened the PermaplanT instance) it doesn't work, because the service worker will always intercept and try to open it within the PWA. Maybe we have to exclude `/api/` and `/.commit` somewhere in Vite's config?
Maybe someone from the frontend team could have a look into this matter.

**Comment by markus2330** [comment]
> I added the Cache-Control: no-cache header. (this should force the use of the ETag header instead of going directly into the cache)

This sounds like a very invasive change. I assume it is only for pr.libelektra.org? Please document.

> Unfortunately, if I want to open it in the browser (after I have already opened the PermaplanT instance) it doesn't work

I don't think this would be a problem, we shouldn't spend time on this. (Btw. for me it works.)

**Comment by markus2330** [comment]
@fel115 can you maybe have a look at this?

**Comment by 4ydan** [comment]
Maybe we can do something like [this](https://doc.rust-lang.org/cargo/guide/cargo-home.html#caching-the-cargo-home-in-ci)?

**Comment by markus2330** [comment]
Yes, that sounds fantastic.

**Comment by 4ydan** [comment]
What do we think about skipping the backend stage if no changes have happened in the backend folder?

**Comment by markus2330** [comment]
This is a **fantastic** idea!

**Comment by markus2330** [comment]
I updated the top post:

- [ ] introduce cargo registry cache
- [ ] completely skip building backend if no changes

**Comment by markus2330** [comment]
@4ydan what about the idea `Completely skip building frontend/backend if no changes` is this a worthwhile idea? Or is it difficult to get the old artifacts from previous builds?

**Comment by 4ydan** [comment]
> Or is it difficult to get the old artifacts from previous builds?

Exactly. Maybe it is possbile to take the artifacts from masters last build?

**Comment by markus2330** [comment]
As everyone is happy now with the speed, let us keep this for later.
